Parallelize RING package tests

moki1202 opened the pull request:
Signed-off-by: Shashank Singh [shashanksingh819@gmail.com](mailto:shashanksingh819@gmail.com)

**What this PR does**:
Optimizes go tests in `ring` package 

**Which issue(s) this PR fixes**:
Partially Fixes #3922 

Previous:
```
go test ./pkg/ring/ -v -count 1 -parallel 1
ok  	github.com/cortexproject/cortex/pkg/ring 71.475s
```
after:
```
go test ./pkg/ring/ -v -count 1
ok  	github.com/cortexproject/cortex/pkg/ring	10.188s

```
**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

moki1202 commented:
@friedrichg tried my best, please let me know if this PR needs improvement. Wanted to ask a question, is it possible that using t.parallel() in some places might actually increase the already low test times rather than optimizing it further?

friedrichg commented:
Thanks for your contribution.

@moki1202 As a general rule, it will use more cpus and will be faster. The only risk is we can make tests flaky if there is hidden bad concurrency in cortex.

Always sign your commits so it can pass the DCO. See [details](https://github.com/cortexproject/cortex/pull/5151/checks?check_run_id=11374077795)

friedrichg commented:
Please look at the lint errors and fix them

moki1202 commented:
**On my machine**-
after-
```
go test -tags netgo -timeout 30m -race -count 1 ./pkg/ring/...
ok      github.com/cortexproject/cortex/pkg/ring        31.153s
```

friedrichg reviewed (changes requested):
Awesome job! 

You are almost there, there is a few missing.

Following files still need some t.Parallel after t.Run

pkg/ring/kv/dynamodb/dynamodb_test.go
pkg/ring/kv/memberlist/memberlist_client_test.go
pkg/ring/model_test.go
pkg/ring/replication_set_tracker_test.go
pkg/ring/replication_strategy_test.go
pkg/ring/ring_test.go

moki1202 commented:
> Awesome job!
> 
> You are almost there, there is a few missing.
> 
> Following files still need some t.Parallel after t.Run
> 
> pkg/ring/kv/dynamodb/dynamodb_test.go pkg/ring/kv/memberlist/memberlist_client_test.go pkg/ring/model_test.go pkg/ring/replication_set_tracker_test.go pkg/ring/replication_strategy_test.go pkg/ring/ring_test.go

Done! Added t.parallel() to t.run() functions that I missed. Also added comment where parallelization wouldn't have worked.

friedrichg reviewed (approved):
Thanks for your patience. 

I noticed you skipped pkg/ring/kv/dynamodb/client_test.go too, but just add comments there, I don't think we can parallelize there too much.

Approved with minor nits.

moki1202 commented:
> Thanks for your patience.
> 
> I noticed you skipped pkg/ring/kv/dynamodb/client_test.go too, but just add comments there, I don't think we can parallelize there too much.
> 
> Approved with minor nits.

@friedrichg This is wierd, thought I'd covered every file. Will look into this again!

moki1202 commented on pkg/ring/model_test.go:
```diff
@@ -8,8 +8,7 @@ import (
 func TestInstanceDesc_IsHealthy_ForIngesterOperations(t *testing.T) {
-	t.Parallel()
-
+	//parallel testing causes data race
```
@friedrichg Will check this again. Now I understand why you emphasized on creating smaller PRs 😅. Much easier to work with, that way.
